src/gevent/_hub_primitives.py

```diff
@@ -133,40 +130,49 @@ def __iter__(self):
     def __next__(self):
         self._begin()
 
-        if self._count == 0:
+        if not self._objects or self._count == 0:
             # Exhausted
-            self._cleanup()
-            raise StopIteration()
+            self.__exit__(None, None, None)
+            raise StopIteration
 
-        self._count -= 1
         try:
             item = self._waiter.get()
             self._waiter.clear()
             if item is self:
                 # Timer expired, no more
-                self._cleanup()
-                raise StopIteration()
+                self.__exit__(None, None, None)
+                raise StopIteration
+
+            self._count -= 1
+            self._objects.remove(item)
+            self._unlink(item)
```

Comment:
Unlinking objects as they become ready breaks the use case that *deliberately* resets them and lets them become ready, up to *count* number of times (think of queue processors). Please restore.